cache/cache.go

```diff
@@ -0,0 +1,190 @@
+package cache
+
+import (
+	"container/list"
+	"sync"
+)
+
+// Cache is a bounded-size in-memory cache of sized items with a configurable eviction policy
+type Cache interface {
+	// Get retrieves items from the cache by key.
+	// If an item for a particular key is not found, its position in the result will be nil.
+	Get(keys ...string) []Item
+
+	// Put adds an item to the cache.
+	Put(key string, item Item)
+
+	// Remove clears items with the given keys from the cache
+	Remove(keys ...string)
+
+	// Size returns the size of all items currently in the cache.
+	Size() uint64
+}
+
+// Item is an item in a cache
+type Item interface {
+	// Size returns the item's size, in bytes
+	Size() uint64
+}
+
+// A tuple tracking a cached item and a reference to its node in the eviction list
+type cached struct {
+	item    Item
+	element *list.Element
+}
+
+// Sets the provided list element on the cached item if it is not nil
+func (c *cached) setElementIfNotNil(element *list.Element) {
+	if element != nil {
+		c.element = element
+	}
+}
+
+// Private cache implementation
+type cache struct {
+	sync.Mutex                                  // Lock for synchronizing Get, Put, Remove
+	cap          uint64                         // Capacity bound
+	size         uint64                         // Cumulative size
+	items        map[string]*cached             // Map from keys to cached items
+	keyList      *list.List                     // List of cached items in order of increasing evictability
```

Comment:
I like "evictability" as a trait. I think I had some floor-mates in college with a really high score in that...